src/System.Private.CoreLib/shared/System/SpanHelpers.Byte.cs

```diff
@@ -959,55 +960,124 @@ public static unsafe int LastIndexOfAny(ref byte searchSpace, byte value0, byte
         public static unsafe bool SequenceEqual(ref byte first, ref byte second, nuint length)
         {
             if (Unsafe.AreSame(ref first, ref second))
+            {
                 goto Equal;
+            }
 
-            IntPtr i = (IntPtr)0; // Use IntPtr for arithmetic to avoid unnecessary 64->32->64 truncations
-            IntPtr n = (IntPtr)(void*)length;
-
-            if (Vector.IsHardwareAccelerated && (byte*)n >= (byte*)Vector<byte>.Count)
+            IntPtr offset = (IntPtr)0; // Use IntPtr for arithmetic to avoid unnecessary 64->32->64 truncations
+            IntPtr nLength = (IntPtr)(void*)length;
+            bool result = false;
+            // Use Avx2.IsSupported check to remove this branch at Jit time if the CPU does not support it.
+            if (Avx2.IsSupported)
             {
-                n -= Vector<byte>.Count;
-                while ((byte*)n > (byte*)i)
+                while ((byte*)nLength >= (byte*)Vector256<byte>.Count)
                 {
-                    if (Unsafe.ReadUnaligned<Vector<byte>>(ref Unsafe.AddByteOffset(ref first, i)) !=
-                        Unsafe.ReadUnaligned<Vector<byte>>(ref Unsafe.AddByteOffset(ref second, i)))
+                    nLength -= Vector256<byte>.Count;
+
+                    Vector256<byte> left = Unsafe.ReadUnaligned<Vector256<byte>>(ref Unsafe.AddByteOffset(ref first, offset));
+                    Vector256<byte> right = Unsafe.ReadUnaligned<Vector256<byte>>(ref Unsafe.AddByteOffset(ref second, offset));
+
+                    offset += Vector256<byte>.Count;
+
+                    if (Avx2.MoveMask(Avx2.CompareEqual(left, right)) != -1)
+                    {
+                        goto NotEqual;
+                    }
+                }
+
+                if ((byte*)nLength >= (byte*)Vector128<byte>.Count)
+                {
+                    nLength -= Vector128<byte>.Count;
+
+                    Vector128<byte> left = Unsafe.ReadUnaligned<Vector128<byte>>(ref Unsafe.AddByteOffset(ref first, offset));
+                    Vector128<byte> right = Unsafe.ReadUnaligned<Vector128<byte>>(ref Unsafe.AddByteOffset(ref second, offset));
+
+                    offset += Vector128<byte>.Count;
+
+                    if (Sse2.MoveMask(Sse2.CompareEqual(left, right)) != 0xFFFF)
                     {
                         goto NotEqual;
                     }
-                    i += Vector<byte>.Count;
                 }
-                return Unsafe.ReadUnaligned<Vector<byte>>(ref Unsafe.AddByteOffset(ref first, n)) ==
-                       Unsafe.ReadUnaligned<Vector<byte>>(ref Unsafe.AddByteOffset(ref second, n));
             }
+            // Use Sse2.IsSupported check to remove this branch at Jit time if the CPU does not support it.
+            else if (!Avx2.IsSupported && Sse2.IsSupported)
```

Comment:
Is the redundant `!Avx2.IsSupported` needed here? I would expect that the `if (Avx2.IsSupported)` above will be allow JIT to dead-code eliminate the else part.